### DISCUSSION TITLE
Next.js API routes with Koa.js

### DISCUSSION THREAD

**Discussion by ivandotv** [vercel/next.js#36941] [Show and tell]
Hi all,

I've been doing some CRUD with Next.js API routes, and I've always liked
using Koa.js for creating the API endpoints in Node.js

I think Koa.js is a great fit for Next.js because it is fully async, error handling is easy, and the codebase is very small so the initialization of the Koa App is very fast.

I've decided to integrate Koa.js (and Koa Router) with the Next.js API routes:

Usage is simple as this (in your API file):

```ts
//pages/api/[[...demo]].ts
import { KoaApi, withKoaApi } from 'nextjs-koa-api'

const api = new KoaApi({ router: { prefix: '/api' } })

api.use((ctx) => {
  ctx.body = 'Hello World'
})
.router.get('/:todo',....)
.post('/:todo',....)
.delete('/:todo',....)

//use helper function
export default withKoaApi(API)

//or the standard way
export default function handler(req: NextApiRequest, res: NextApiResponse) {
  return api.run(req, res)
}

```

Check out the [repository](https://github.com/ivandotv/nextjs-koa-api), where you can find complete documentation and a small Next.js demo app.

I'm open to suggestions and contributions.

Thanks!

**Comment by vanduc1102** [comment]
I am using this code from chatgpt:

```ts
// pages/api/third-party/[...all].ts
// nextjs@13, koa@2.15

import type { NextApiRequest, NextApiResponse } from "next";

import { createServer } from "http";
import Koa, { Context } from "koa";

const koaApp = new Koa();

koaApp.use(async (ctx: Context) => {
  ctx.body = "Hello from Koa.js in Next.js API route!";
});

const server = createServer(koaApp.callback());

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  server.emit("request", req, res);
}

```
